### DISCUSSION TITLE
Struggling with figure sizes, error bars and axis scales with Matplotlib

### DISCUSSION THREAD

**Discussion by benpymer** [Part-II-Computational-Physics/.github#36] [General]
Hi there, I expect these questions are pretty simple, but I'm struggling to find helpful things online.

Firstly, I've written a little loop for task 3 such that it loops through the different Ns values and prints out a graph for each one all at once in one subplot, however I have just used the plt.subplot(rows, columns, index) command to create the subplot, without bothering with all the fig, ax stuff. However all of my plots and their axes have ended up printing on top of each other, and the plots only occupy about half the page. I was just wondering if there was an easy way to resize the plots and spread them out so that they aren't overlapping?

I'm also doing task 4 right now. My code to produce the plot is this

plt.errorbar(Ns, [time.average for time in timingCompact], yerr = [time.stdev for time in timingCompact])
plt.yscale('log')

For some reason, matplotlib doesn't like the plt.yscale('log') command, despite being very happy to take a plt.xscale('log') command in task 3. Can you not rescale the axis using plt.errorbar? If not, how is best to add the errorbars? If it's some other problem, any idea what?

Thanks, Ben

**Comment by JNHallen** [comment]
I would suggest plotting the results for different $N$ in task 3 on top of each other, rather than in different subplots. Partly because it makes it easier to compare the results, and partly because having loads of subplots in a jupyter notebook is a bit messy. (You will want to normalise the energies by dividing them by $N$ if you are doing this.) 

For your issue with log-scales in task 4, the most likely issue is that you have values where the average - standard deviation < 0. That would throw an error. If that is not the issue you can perhaps include the error message here and I will take a look. 


If you really want to use subplots, you can do something like:
```
hh = np.logspace(-2, 2, 10)
Ns = np.arange(2, 11, 1)

fig, axs = plt.subplots(3, 3)

for ii, N in enumerate(Ns):
    Groundstates = []
    for h in hh:
        Groundstates.append(...)
    

    axs[ii//3, ii%3].plot(hh, Groundstates, '-o')
```

Alternatively you can just plot all the results in separate figures by including "plt.figure(N)" in your loop.

**Reply by benpymer** [reply]
Hi, thanks so much!

I have now changed the code to plot all the different N values on the same graph. Unfortunately I'm now getting the error message for both tasks. The message I get is "'str' object is not callable". I tried deleting the ' ' either side of the log and symlog to make it not a string, but it then said "name 'symlog/log' is not defined". Am I just making a syntax error somewhere?

**Reply by JNHallen** [reply]
Hi! If you are getting that error message, that means you are trying to use a string object as if it was something callable. A basic example would be if you import matplotlib.pyplot as plt, but then define a string plt="SomeString". If you then try doing something like plt.plot(...) you will get that error message. Without seeing your code I cannot tell you exactly what has gone wrong though.

**Reply by benpymer** [reply]
Turns out it was a simple case of restarting the kernel and trying again. Turning it off and back on again never fails!
